packages/plugin-cloudflare/src/tunnel.ts

```diff
@@ -10,61 +11,67 @@ import {fileURLToPath} from 'url'
 
-export async function hookStart(port: number): ReturnType {
+let currentStatus: TunnelStatusType = {status: 'not-started'}
+
+export const getCurrentStatus = async () => currentStatus
+
+export const abortController = new AbortController()
+
+export async function hookStart(port: number) {
   try {
-    const {url} = await tunnel({port})
-    return ok({url})
+    return tunnel({port})
     // eslint-disable-next-line no-catch-all/no-catch-all, @typescript-eslint/no-explicit-any
   } catch (error: any) {
     const tunnelError = new TunnelError('unknown', error.message)
     return err(tunnelError)
   }
 }
 
-async function tunnel(options: {port: number}): Promise<{url: string}> {
+function tunnel(options: {port: number}): void {
   const args: string[] = ['tunnel', '--url', `http://localhost:${options.port}`, '--no-autoupdate']
   const errors: string[] = []
 
   let connected = false
   let resolved = false
   let url: string | undefined
+  currentStatus = {status: 'starting'}
 
-  return new Promise<{url: string}>((resolve, reject) => {
-    setTimeout(() => {
-      if (!resolved) {
-        resolved = true
-        const lastErrors = errors.slice(-5).join('\n')
-        reject(new Error(`Timed out while creating a cloudflare tunnel: ${lastErrors}`))
-      }
-    }, TUNNEL_TIMEOUT * 1000)
+  setTimeout(() => {
+    if (!resolved) {
+      resolved = true
+      const lastErrors = errors.slice(-5).join('\n')
+      currentStatus = {status: 'error', message: lastErrors}
+    }
+  }, TUNNEL_TIMEOUT * 1000)
 
-    const customStdout = new Writable({
-      write(chunk, _, callback) {
-        outputDebug(chunk.toString())
-        if (resolved) return
-        if (!url) url = findUrl(chunk)
-        if (findConnection(chunk)) connected = true
-        if (connected) {
-          if (!url) return reject(new Error('A connection was established but no Tunnel URL was found'))
+  const customStdout = new Writable({
+    write(chunk, _, callback) {
+      outputDebug(chunk.toString())
+      if (resolved) return
+      if (!url) url = findUrl(chunk)
+      if (findConnection(chunk)) connected = true
+      if (connected) {
+        if (url) {
           resolved = true
-          resolve({url})
+          currentStatus = {status: 'connected', url, port: options.port}
+        } else {
+          currentStatus = {status: 'error', message: 'Could not find tunnel url'}
         }
-        const errorMessage = findError(chunk)
-        if (errorMessage) errors.push(errorMessage)
-        callback()
-      },
-    })
+      }
+      const errorMessage = findError(chunk)
+      if (errorMessage) errors.push(errorMessage)
+      callback()
+    },
+  })
 
-    // eslint-disable-next-line @typescript-eslint/no-floating-promises
-    exec(getBinPathTarget(), args, {
-      stdout: customStdout,
-      stderr: customStdout,
-    })
+  // eslint-disable-next-line @typescript-eslint/no-floating-promises
+  exec(getBinPathTarget(), args, {
```

Comment:
We should test this with `exit 1` -- i.e. what happens if the commands fails quickly (vs. it eventually times out, which I think we do cover for already)

Reply:
Added an automatic retry if the process crashes before getting a valid URL and connection.

Once we get a valid URL and the connection is stablished, if it crashes we can't do anything, just let it crash with a better message. (because the CLI is already using that URL and can't tell it to use a new one)